File: engine/core/ai.py
import os
import httpx
import json
import logging
import asyncio

logger = logging.getLogger("contapyme.ai")

# Configuración de Groq Cloud (v8.7 Smart Cloud)
GROQ_URL = "https://api.groq.com/openai/v1/chat/completions"
GROQ_API_KEY = os.getenv("GROQ_API_KEY", "")
DEFAULT_MODEL = "openai/gpt-oss-120b"
FALLBACK_MODEL = "openai/gpt-oss-20b"
TERTIARY_MODEL = "qwen/qwen3.6-27b"

async def groq_chat_completion(messages: list, temperature: float = 0.0, json_mode: bool = True) -> str | None:
    """
    Función genérica para ejecutar inferencia en Groq Cloud con reintentos y fallbacks.
    """
    if not GROQ_API_KEY:
        logger.error("[AI] ⚠️ Falta GROQ_API_KEY. Abortando llamada a Groq.")
        return None

    payload = {
        "model": DEFAULT_MODEL,
        "messages": messages,
        "temperature": temperature,
    }
    if json_mode:
        payload["response_format"] = {"type": "json_object"}

    models_to_try = [DEFAULT_MODEL, FALLBACK_MODEL, TERTIARY_MODEL]
    headers = {
        "Authorization": f"Bearer {GROQ_API_KEY}",
        "Content-Type": "application/json"
    }

    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            for current_model in models_to_try:
                payload["model"] = current_model
                try:
                    resp = await client.post(GROQ_URL, json=payload, headers=headers)
                    if resp.status_code == 200:
                        data = resp.json()
                        return data.get("choices", [{}])[0].get("message", {}).get("content", "")
                    if resp.status_code == 429:
                        logger.warning(f"[AI] ⏱️ Rate limit en {current_model}. Probando siguiente modelo...")
                        await asyncio.sleep(1.5)
                    else:
                        logger.warning(f"[AI] ⚠️ Groq {resp.status_code} con {current_model}...")
                except Exception as ex:
                    logger.warning(f"[AI] Excepción en modelo {current_model}: {ex}")
    except Exception as e:
        logger.error(f"[AI] Error ejecutando groq_chat_completion: {e}")

    return None


async def process_news_with_local_llm(headline: str, content: str = "") -> dict:
    """
    Procesa una noticia usando Groq Cloud para resumirla,
    darle un tono ejecutivo y verificar su relevancia regional.
    """
    if not GROQ_API_KEY:
        logger.error("[AI] ⚠️ Falta GROQ_API_KEY. Abortando procesamiento de IA.")
        return None

    prompt = f"""
    Eres el editor y Analista Principal de 'Contapymepuq', un portal institucional y financiero de noticias de Magallanes, Chile.
    Tu audiencia son contadores, directores de empresas y profesionales que exigen rigor, veracidad y análisis estratégico.
    Tu misión es transformar hechos noticiosos en artículos de alta calidad periodística, con redacción propia y original (evitando cualquier plagio o copia textual), pero con una FIDELIDAD FACTUAL ABSOLUTA AL 100% (cero alucinaciones).
    
    ¡FILTRO DE RELEVANCIA!: Si la noticia es sobre farándula, chismes, curiosidades mundiales virales o temas que no afectan a la economía, al derecho o a la vida en Magallanes, NO LA REDACTES. Responde con un JSON que tenga "category": "IGNORE".

    TEXTO BASE CAPTURADO:
    Titular de referencia: {headline}
    Hechos reportados: {str(content)[0:3500]}

    REGLAS DE ORO EDITORIALES (BLINDAJE FACTUAL Y ANTI-PLAGIO):
    1. REESCRITURA ORIGINAL (CERO PLAGIO): Redacta la noticia con tu propia estructura sintáctica y vocabulario ejecutivo. No copies frases completas de la fuente. Transforma la información en una pieza periodística original de Contapymepuq.
    2. FIDELIDAD FACTUAL ESTRICTA (CERO ALUCINACIÓN): 
       - Cíñete ÚNICA Y EXCLUSIVAMENTE a los hechos, cifras, lugares, nombres y procedimientos descritos en el texto base.
       - NUNCA inventes, supongas ni extrapoles argumentos de defensa, descargos de abogados, motivos personales ni justificaciones de las partes involucradas si no están explícitamente detallados en el texto.
       - Si la noticia menciona una investigación o denuncia (ej. Aduanas, PDI, SII, Fiscalía), expón los hechos objetivos de forma neutral, sin calificar ni inventar explicaciones de los acusados o instituciones.
    3. LONGITUD PROPORCIONAL Y SIN RELLENO: Desarrolla el texto de manera concisa y sustanciosa, proporcional a la cantidad de información real del texto base. No agregues párrafos vacíos o redundantes para inflar el texto.
    4. TONO: Ejecutivo, sobrio, formal e institucional.
    5. ESTRUCTURA REQUERIDA:
       - 'title': Titular profesional (MÁX. 10 PALABRAS). No usar mayúsculas sostenidas.
       - 'summary': Resumen ejecutivo directo de 2 a 3 oraciones con los puntos clave.
       - 'full_content': Redacción fluida y completa dividida en párrafos bien estructurados con análisis del contexto regional/económico cuando aplique.
    6. CATEGORÍAS PERMITIDAS: INVERSIONES, ECONOMÍA, FINANZAS, SII/LEGAL, MAGALLANES ACTUAL, DEPORTES REGIONALES.
    7. SEGURIDAD JSON: NUNCA uses comillas dobles (") dentro de los valores de texto. Si necesitas citar algo, usa comillas simples (').

    RESPONDE EXCLUSIVAMENTE EN FORMATO JSON:
    {{
        "title": "Titular reescrito profesional",
        "category": "CATEGORÍA",
        "summary": "Resumen ejecutivo directo.",
        "full_content": "Cuerpo completo de la noticia.",
        "is_featured": boolean,
        "brand_name": "Nombre de la marca o entidad involucrada si aplica (ej: EDELMAG, ENAP, SII), de lo contrario null",
        "visual_prompt": "Detailed visual description in English for authentic editorial press photography in Magallanes, Chile (e.g. 'Documentary press photo of cargo trucks in Punta Arenas port, overcast Magallanes cold sky, shot on 35mm lens, natural daylight'). Do NOT include cartoon, 3d, or CGI terms.",
        "seo_keywords": "Palabras clave separadas por comas (máximo 5) relevantes para SEO.",
        "seo_description": "Meta descripción breve de menos de 160 caracteres."
    }}
    """

    payload = {
        "model": DEFAULT_MODEL,
        "messages": [{"role": "user", "content": prompt}],
        "temperature": 0.0,
        "response_format": {"type": "json_object"}
    }

    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            headers = {
                "Authorization": f"Bearer {GROQ_API_KEY}",
                "Content-Type": "application/json"
            }
            
            models_to_try = [DEFAULT_MODEL, FALLBACK_MODEL, TERTIARY_MODEL]
            response = None
            
            for current_model in models_to_try:
                payload["model"] = current_model
                try:
                    response = await client.post(GROQ_URL, json=payload, headers=headers)
                    if response.status_code == 200:
                        break
                    
                    if response.status_code == 429:
                        # Si hay límite de tasa, esperar brevemente
                        logger.warning(f"[AI] ⏱️ Rate limit en {current_model}. Probando siguiente modelo...")
                        await asyncio.sleep(2.0)
                    else:
                        logger.warning(f"[AI] ⚠️ Groq {response.status_code} con {current_model}. Probando siguiente modelo...")
                except Exception as req_ex:
                    logger.warning(f"[AI] Error solicitando modelo {current_model}: {req_ex}")
            
            if response and response.status_code == 200:
                result = response.json()
                content_raw = result.get("choices", [{}])[0].get("message", {}).get("content", "{}")
                data = json.loads(content_raw)
                
                if data.get("category") == "IGNORE":
                    logger.info(f"[AI] Noticia ignorada por irrelevancia: {headline}")
                    return None

                return {
                    "title": data.get("title", headline),
                    "category": data.get("category", "REGIONAL"),
                    "summary": data.get("summary", ""),
                    "full_content": data.get("full_content", content),
                    "is_featured": data.get("is_featured", False),
                    "brand_name": data.get("brand_name", None),
                    "visual_prompt": data.get("visual_prompt", f"A hyperrealistic news photograph of Punta Arenas, Magallanes, Patagonia, 8k, cinematic lighting, realistic colors."),
                    "seo_keywords": data.get("seo_keywords", "contapymepuq, magallanes, punta arenas, chile"),
                    "seo_description": data.get("seo_description", data.get("summary", "")[:155])
                }
            else:
                logger.error(f"[AI] Error en Groq: {response.status_code} - {response.text}")
    except Exception as e:
        logger.error(f"[AI] No se pudo procesar con IA: {e}")
    
    return None

File: app/src/actions/ai.ts
'use server'

import { createClient } from '@/lib/supabase/server'
import { checkAdminPermission } from './news'

const GROQ_URL = 'https://api.groq.com/openai/v1/chat/completions'
const DEFAULT_MODEL = 'llama-3.3-70b-versatile'

interface AssistInput {
  draftTitle?: string
  draftContent: string
  draftCategory?: string
}

interface AssistResult {
  success: boolean
  error?: string
  data?: {
    title: string
    category: string
    summary: string
    content: string
  }
}

/**
 * Server Action que utiliza Groq Cloud AI para pulir, redactar de forma experta
 * y estructurar un borrador de noticia ingresado por el usuario.
 */
export async function assistNewsWritingAction(input: AssistInput): Promise<AssistResult> {
  const apiKey = process.env.GROQ_API_KEY || ''

  try {
    // 1. Validar que el usuario sea administrador
    const authCheck = await checkAdminPermission()
    if (!authCheck.authorized) {
      return { success: false, error: 'No autorizado: requiere privilegios editoriales' }
    }

    if (!apiKey) {
      console.error('[AI Assist] GROQ_API_KEY no configurado en el servidor Next.js')
      return { 
        success: false, 
        error: 'El servicio de asistencia por IA no está disponible temporalmente (Motivo: GROQ_API_KEY vacía o no detectada)' 
      }
    }

    if (!input.draftContent || input.draftContent.trim().length < 10) {
      return { success: false, error: 'El texto borrador debe tener al menos 10 caracteres' }
    }

    // 2. Armar el Prompt enfocado en el estilo editorial de Contapymepuq
    const systemPrompt = `
    Eres el redactor jefe e Inteligencia Artificial editorial de 'Contapymepuq', el principal diario digital de Punta Arenas y la región de Magallanes, Chile.
    Tu audiencia son contadores, gerentes, empresarios y profesionales de la Patagonia que buscan información seria y estructurada.

    TAREA:
    Tu tarea es transformar el borrador o apuntes crudos ingresados por el usuario en un artículo de prensa profesional y pulido de nivel periodístico.

    REGLAS EDITORIALES DE ORO:
    1. TONO: Sobrio, formal, analítico y altamente periodístico. Evita sensacionalismos y clickbaits.
    2. ENFOQUE REGIONAL: Si el borrador menciona a Punta Arenas, Magallanes, Tierra del Fuego, Puerto Natales o aspectos locales de la Patagonia, dale un realce especial al impacto local.
    3. ESTRUCTURA:
       - 'title': Titular profesional (máximo 10 palabras), con mayúsculas y minúsculas adecuadas (no todo mayúsculas). Debe ser directo y descriptivo.
       - 'summary': Resumen ejecutivo en un solo párrafo corto (máximo 3 líneas) enfocado en lo esencial.
       - 'content': Cuerpo completo de la noticia, redactado de forma experta en 3 o 4 párrafos bien estructurados (mínimo 200 palabras en total).
    4. CATEGORÍA: Clasifica el artículo estrictamente en una de las siguientes 3 categorías oficiales:
       - 'MAGALLANES ACTUAL' (Para noticias locales, comunidad, clima, deportes, eventos, etc.)
       - 'ECONOMÍA' (Para noticias financieras, mercado, inversiones, PYMEs, IPC, dólar, etc.)
       - 'SII / LEGAL' (Para noticias de impuestos, circulares del SII, leyes tributarias, laboral, etc.)
    5. SEGURIDAD DE FORMATO: NUNCA uses comillas dobles (") dentro de las respuestas de texto. Si necesitas citar algo, usa comillas simples ('). Esto es extremadamente crítico para que el formato JSON de respuesta no se rompa al deserializar.

    RESPONDE EXCLUSIVAMENTE EN FORMATO JSON VÁLIDO CON LA SIGUIENTE ESTRUCTURA:
    {
      "title": "Titular profesional del artículo",
      "category": "MAGALLANES ACTUAL o ECONOMÍA o SII / LEGAL",
      "summary": "Resumen ejecutivo del artículo...",
      "content": "Cuerpo completo del artículo en párrafos expertos..."
    }
    `

    const userPrompt = `
    BORRADOR O APUNTES DEL USUARIO:
    Título tentativo: ${input.draftTitle || 'Sin título'}
    Categoría elegida: ${input.draftCategory || 'Sin categoría fija'}
    Contenido borrador:
    """
    ${input.draftContent}
    """
    `

    // 3. Consultar a la API de Groq Cloud
    const response = await fetch(GROQ_URL, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        model: DEFAULT_MODEL,
        messages: [
          { role: 'system', content: systemPrompt },
          { role: 'user', content: userPrompt }
        ],
        temperature: 0.3,
        response_format: { type: 'json_object' }
      })
    })

    if (!response.ok) {
      const errText = await response.text()
      console.error(`[AI Assist] Error de API de Groq (${response.status}):`, errText)
      return { 
        success: false, 
        error: `Error de API de Groq (Status: ${response.status}): ${errText.substring(0, 150)}` 
      }
    }

    const result = await response.json()
    const rawContent = result.choices?.[0]?.message?.content || '{}'
    const parsedData = JSON.parse(rawContent)

    return {
      success: true,
      data: {
        title: parsedData.title || input.draftTitle || 'Artículo Asistido por IA',
        category: parsedData.category || input.draftCategory || 'MAGALLANES ACTUAL',
        summary: parsedData.summary || '',
        content: parsedData.content || input.draftContent
      }
    }

  } catch (err: any) {
    console.error('[AI Assist] Error crítico en Server Action:', err.message)
    return { success: false, error: `Error crítico en la asistencia por IA (Diagnóstico: ${err.message})` }
  }
}
